android/app/src/main/java/org/jetbrains/kotlinconf/ui/SessionDetailsFragment.kt

```diff
@@ -38,85 +53,131 @@ class SessionDetailsFragment : Fragment(), SessionDetailsView {
-        badButton.setOnClickListener { presenter.rateSession(SessionRating.BAD) }
-    }
 
-    override fun onDestroy() {
-        super.onDestroy()
-        presenter.onDestroy()
-    }
+        setHasOptionsMenu(true)
+
+        (activity as AppCompatActivity).apply {
+            setSupportActionBar(toolbar)
+            supportActionBar?.setDisplayShowHomeEnabled(true)
+            supportActionBar?.setDisplayShowTitleEnabled(false)
+            supportActionBar?.setDisplayHomeAsUpEnabled(true)
+        }
+
+        val sessionId = arguments?.get(KEY_SESSION_ID) as String
+        val viewModel = ViewModelProviders.of(
+            this,
+            AndroidViewModelFactory.getInstance(activity!!.application)
+        )
+            .get(SessionDetailsViewModel::class.java)
+            .apply { setSession(sessionId) }
+
+        favoriteButton.setOnClickListener {
+            launch(UI) { viewModel.toggleFavorite() }
+        }
+
+        val clickListener = View.OnClickListener { view: View ->
+            val rating = when (view) {
+                goodButton -> SessionRating.GOOD
+                okButton -> SessionRating.OK
+                badButton -> SessionRating.BAD
+                else -> null
+            }
+
+            launch(UI) {
+                if (rating != null) {
+                    if (viewModel.rating.value != rating) {
+                        viewModel.setRating(rating)
+                    } else {
+                        viewModel.removeRating()
+                    }
+                }
+            }
+        }
 
-    override fun setIsFavorite(isFavorite: Boolean) {
-        val icon = if (isFavorite) R.drawable.ic_favorite_white_24dp else R.drawable.ic_favorite_border_white_24dp
-        favoriteButton.setImageResource(icon)
+        viewModel.session.observe(this, this::updateView)
+
+        viewModel.isFavorite.observe(this) { isFavorite ->
+            if (isFavorite == true) {
+                favoriteButton.setImageResource(R.drawable.ic_favorite_white_24dp)
+            } else {
+                favoriteButton.setImageResource(R.drawable.ic_favorite_border_white_24dp)
+            }
+        }
+
+        setupRatingButtons(viewModel.rating.value)
+
+        viewModel.rating.observe(this) { rating ->
+            setupRatingButtons(rating)
+        }
+
+        goodButton.setOnClickListener(clickListener)
+        okButton.setOnClickListener(clickListener)
+        badButton.setOnClickListener(clickListener)
     }
 
     override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
         super.onCreateOptionsMenu(menu, inflater)
         menu.clear()
     }
 
-    override fun setRatingClickable(clickable: Boolean) {
-        goodButton.isClickable = clickable
-        okButton.isClickable = clickable
-        badButton.isClickable = clickable
+    private fun setupRatingButtons(rating: SessionRating?) {
+        goodButton.backgroundResource = if (rating == SessionRating.GOOD)
```

Comment:
If we extract the `if-else`:
```kotlin
if (condition)
    R.drawable.round_toggle_button_background_selected
else
    R.drawable.round_toggle_button_background
```
to a separate method, we could remove the duplication in setting the background. But considering the size of the PR and how important it is to be merged, it's definitely a nitpick. I'd recommend merging the PR first and do some cleaning later on.